plugins/uefi/fu-self-test.c

```diff
@@ -44,18 +46,18 @@ static void
 fu_uefi_pcrs_2_0_func (void)
 {
 	g_autofree gchar *devpath = NULL;
-	g_autofree gchar *sysfstpmdir = NULL;
 	const gchar *tpm_server_running = g_getenv ("TPM_SERVER_RUNNING");
 	gboolean ret;
 	g_autoptr(FuUefiPcrs) pcrs = fu_uefi_pcrs_new ();
 	g_autoptr(GError) error = NULL;
 	g_autoptr(GPtrArray) pcr0s = NULL;
 	g_autoptr(GPtrArray) pcrXs = NULL;
+	gboolean physical_tpm;
 
-	sysfstpmdir = fu_common_get_path (FU_PATH_KIND_SYSFSDIR_TPM);
-	devpath = g_build_filename (sysfstpmdir, "tpm0", NULL);
-
-	if (!g_file_test (devpath, G_FILE_TEST_EXISTS) && (tpm_server_running == NULL)) {
+	devpath = g_build_filename ("/dev", "tpm0", NULL);
+	physical_tpm = g_file_test (devpath, G_FILE_TEST_EXISTS) &&
+		       g_access (devpath, W_OK | R_OK) == 0;
```

Comment:
that works, but it's a bit of a cludge. can't we do a `g_error_matches()` on the error instead?

Reply:
Can't we map the output of `Esys_Initialize()` to differing GError codes? At the moment they're all mapped to `G_IO_ERROR_NOT_SUPPORTED`

Reply:
When the second argument of `Esys_Initialize` is `NULL`, [different libraries](https://github.com/tpm2-software/tpm2-tss/blob/master/src/tss2-tcti/tctildr-dl.c#L33-L54) are tried until one that works is found. Its return value is `TSS2_ESYS_RC_SUCCESS` if everything worked as expected, or `TSS2_TCTI_RC_IO_ERROR` if initialisation failed, no further information on the nature of the error (missing libraries/permission problems/TPM error/...) is available. It is possible to use the second argument to specify through which library the TPM should be accessed, but usually it is best to leave this to tpm2-tss.

Checking whether we have permission to communicate with the TPM is not easy: first of all, both `/dev/tpm0` and `/dev/tpmrm0` can be used to access it, so we would need to check if we have RW permissions for either of these. But even if the current user doesn't have permissions to access these devices, tpm2-abrmd might be installed and running as the `tss` user, which means that we don't have direct access to the TPM at all, but can still talk to it using the resource manager.

To test whether it makes sense to run the test, we could try running `Esys_Initialize` and see whether that succeeds:
```
TSS2_RC rc;
g_autoptr(ESYS_CONTEXT) ctx = NULL;
rc = Esys_Initialize (&ctx, NULL, NULL);
if (rc != TSS2_RC_SUCCESS) {
	g_test_skip ("no physical or simulated TPM 2.0 device available");
	return;
}
```
If that fails, there is definitely something wrong on the tpm2-tss side, so it does not make sense to run the test.

Reply:
I like this version.